tests/connection_state_test.c

```diff
@@ -0,0 +1,947 @@
+/*
+ * Copyright 2010-2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include <aws/mqtt/private/client_impl.h>
+
+#include <aws/io/channel_bootstrap.h>
+#include <aws/io/event_loop.h>
+#include <aws/io/host_resolver.h>
+
+#include <aws/common/clock.h>
+#include <aws/common/condition_variable.h>
+
+#include <aws/testing/aws_test_harness.h>
+
+#include "mqtt_mock_server_handler.h"
+
+#ifdef _WIN32
+#    define LOCAL_SOCK_TEST_PATTERN "\\\\.\\pipe\\testsock%llu"
+#else
+#    define LOCAL_SOCK_TEST_PATTERN "testsock%llu.sock"
+#endif
+
+struct mqtt_connection_state_test {
+    struct aws_allocator *allocator;
+    struct aws_channel *server_channel;
+    struct aws_channel_handler *test_channel_handler;
+    struct aws_client_bootstrap *client_bootstrap;
+    struct aws_server_bootstrap *server_bootstrap;
+    struct aws_event_loop_group el_group;
+    struct aws_host_resolver host_resolver;
+    struct aws_socket_endpoint endpoint;
+    struct aws_socket *listener;
+    struct aws_mqtt_client mqtt_client;
+    struct aws_mqtt_client_connection *mqtt_connection;
+    struct aws_socket_options socket_options;
+    bool session_present;
+    bool connection_completed;
+    bool client_disconnect_completed;
+    bool server_disconnect_completed;
+    bool connection_interupted;
```

Comment:
interrupted

Also, interrupt/resume can happen multiple times, should these be counters? For that matter, should connection_completed, subscribe_completed, and so on also be counters?

Reply:
they get flipped after they're used. They're only used for semaphores so counters wouldn't buy us anything.